[34444] Projects copy APIv3

oliverguenther opened the pull request:
- Form endpoint
  - [x] Docs
  - [X] implementation
  - [x] Basic test
  - [x] Edge cases / meta spec
  - ~~Determine default selected meta flags~~ I decided for now that the API should not magically define which modules to copy. The decision of default associations to copy is random anyway (Forums are not copied by default). In the API, all dependencies are selected for copy unless specified otherwise
  - [x] Extend schema with description using count

- Create endpoint
  - [x] Redirect to jobs status
  - [x] Extract common logic from controller
  - [x] Docs
  - [x] Basic test
  - [x] Edge cases / meta spec


- Other todos
  - [x] Take care of the send notifications param

oliverguenther committed:
Create copy project endpoint

oliverguenther committed:
Add representers

oliverguenther committed:
Add _meta representer for copy module information

oliverguenther committed:
Add Meta payload representer

oliverguenther committed:
Extract parsing of copy attributes into service

oliverguenther committed:
Extract enqueue job for projects copy

oliverguenther committed:
Keep request object available in the bodied endpoint

this allows us to access grape (e.g., for redirecting)

oliverguenther committed:
Add DelayedModify endpoint that redirects to job status

oliverguenther committed:
Use DelayedModify endpoint for copying projects

oliverguenther committed:
Add api paths for form/copy

oliverguenther committed:
Disable cache on ProjectCopyPayload

oliverguenther committed:
Add spec

oliverguenther committed:
Add resource spec

oliverguenther committed:
Extend schema for meta

oliverguenther committed:
Extend docs

oliverguenther committed:
Don't pass service result, but state as meta to forms

oliverguenther committed:
Use copy dependencies for naming copy options in API

oliverguenther committed:
Add description property for counting

oliverguenther committed:
Remove unused action attribute

oliverguenther committed:
Skip writable checks on meta property for payloads

oliverguenther committed:
Use prepend to allow create(...) usage with meta

oliverguenther committed:
Extend copy spec with custom fields usage

oliverguenther committed:
Add spec for copy flags

oliverguenther committed:
Add todos

oliverguenther committed:
Pass errors correctly to copy settings

oliverguenther committed:
Remove invalid parameters to copy that are now caught through project service

oliverguenther committed:
Remove duplicated validation

oliverguenther committed:
Linting

oliverguenther committed:
Make count a human readable, formattable description string

oliverguenther committed:
Make source count string readable

oliverguenther committed:
Provide the source project for generating the counts

oliverguenther committed:
Extract copying of attachments into sepearate dependent services

This will allow the API to dynamically generate a copy association schema from each dependent service

oliverguenther committed:
Add spec for copy schema representer

oliverguenther committed:
Add payload representer spec

oliverguenther committed:
Also validate the model to pass the validations when copying

oliverguenther committed:
Make description optional and add note to schemas.apib

oliverguenther committed:
Add send_notifications to representer

oliverguenther committed:
Allow to configure sendNotifications

oliverguenther committed:
Allow bodied to receive a state process callback

That allows us to manage the meta state on the API level, not on the service level

oliverguenther committed:
Make other services compatible with BaseCallable

oliverguenther committed:
Fix params passing from BaseCallable

Wrapping a single hash into kwargs obviously loses their key indifference
and we can't expect all services to use a single params object. Older services use kwargs

ulferts commented on docs/api/apiv3/endpoints/projects.apib:
```diff
@@ -112,8 +112,9 @@ Depending on custom fields defined for projects, additional properties might exi
-    + id (required, integer, `1`) ... Project id
-
+    
++ id (required, integer, `1`) ... Project id
```
The changes in indentation will cause the parameters to not be listed again.

ulferts commented on docs/api/apiv3/endpoints/projects.apib:
```diff
@@ -1301,6 +1304,516 @@ For more details and all possible responses see the general specification of [Fo
+                    },
+                    "copyVersions": {
+                        "type": "Boolean",
+                        "name": "versions",
```
Typo 'versions' -> 'Versions'. 

For all the copyXYZ attributes, it would make sense to add the copy character to the name, e.g. "Copy version". This might however only be true from an API perspective, not necessarily from the perspective of our client.

oliverguenther replied:
The current version without copy in ints label corresponds to the current labels that are used in the current project form, so I'd probably not change it.

ulferts commented on docs/api/apiv3/endpoints/schemas.apib:
```diff
@@ -40,16 +40,17 @@ Only one of the links (`allowedValues`, `allowedValuesSchemas`) will exist for a
+| location          | If present, contains a reference to the location of the property in the JSON       | String       |         |
+| description       | If present, contains a formattable, human readable description                     | Formattable  |         |
 
 All of the above properties that do not have a default value *must* be present in the schema.
```
The addition of the `_meta` group should also be described in the `forms.apib` documenation as it will be a general pattern that clients should expect to happen in all payload objects.

ulferts commented on lib/api/utilities/endpoints/bodied.rb:
```diff
@@ -91,6 +104,7 @@ def process(current_user, instance, params)
 
           process_service
             .new(**args.compact)
+            .with_state(process_state.call(model: instance, params: params))
```
I have been wondering for some time now whether it would make sense to make the params available within the state of a service. That way, we would not have to pass the params around and could also access them in the callbacks which could then more easily be turned into default `ActiveModel::Callbacks`.

Do you have an opinion on this, @oliverguenther ?

oliverguenther replied:
Yes that sounds like a good convention.

ulferts reviewed (changes requested):
Sorry, pressed the wrong button so therefore the review is in two parts now. Codewise, there is nothing to complain about except some tiny things.

This is what I notice when poking around:

- [x] `"copyXYZ": true` is default for all copyable associations but setting only one `"copyXYZ": false` (or true) will set all to `false`:
![image](https://user-images.githubusercontent.com/617519/115026355-b0ae7780-9ec2-11eb-8164-6e28090dde60.png)
- [x] when not providing an identifier in the body of the form request, a validation error is displayed. But the identifier property is not displayed in the payload unless a value is provided. Probably existed before.
- [x] I assume that the description on the `copyXYZ` fields is too technical to be displayed in the UI client. But we can also change this once this assumption is validated.
![image](https://user-images.githubusercontent.com/617519/115027245-c2dce580-9ec3-11eb-81bc-36f234a53047.png)
- [x] The links of the form resource do not point to the correct end points:
![image](https://user-images.githubusercontent.com/617519/115027455-06375400-9ec4-11eb-8e0d-a6ec74ae7ee3.png)
- [x] For an API client, it would be nice to also have a redirect url when the project is copied. Currently, only an html url is displayed:
![image](https://user-images.githubusercontent.com/617519/115028347-07b54c00-9ec5-11eb-9df6-7bf533c8f8ea.png)

Having that delayed/redirect flow in place in the API now looks rather promising. We have a couple of end points which can now make use of this (DELETE projects, groups, users, ...).

oliverguenther committed:
Copy wiki page with parent_id

This will save some memory instead of memoizing the entire wiki page

oliverguenther committed:
Fix typo

oliverguenther committed:
Default to true for all copy associations

That means if only a true value is passed, all other options will still be true

oliverguenther committed:
Do not try to copy attachments if the base dependency wasnt copied

oliverguenther committed:
Ensure null identifier gets rendered

oliverguenther committed:
Fix paths to the form and commit

oliverguenther committed:
Change redirect to an URL

oliverguenther committed:
Extend documentation on meta properties as table

oliverguenther committed:
Ensure we pass a default params if incoming params are nil

Now that SetAttributes needs to splat params, they do no longer
correctly get the no-param-default behavior of Ruby, so we need to
explictly assign a default params hash

ulferts reviewed (approved):
:+1: looks very good. Fine to merge once the feature spec is adapted.

oliverguenther committed:
Fix expect for trailing path
